### PR TITLE
[EC-319] Clients: Use Jest Mocks

### PR DESCRIPTION
**Pull request by eliykat**
See the proposed ADR in the diff.

### CODE REVIEW

**Comment by Hinton** [comment]
https://github.com/bitwarden/jslib/pull/752

> I looked into how jest proposes to handle mocks since we currently use the substitute library. And by default jest seems to assume you are not using a DI framework which results in all of their mocking solutions being fairly clunky.
>
> With that out of the way, the cleanest way that I have found to interact with the jest mocking was to use the [jest-mock-extended](https://github.com/marchaos/jest-mock-extended) which is a fairly small library that focuses on adding typing to mocks.
>
> The biggest difference in these are jest-mock-extended returns undefined for all methods of the mocked service. While @fluffy-spoon/substitute will return objects if the service returns an object. During my testing this showed up as some methods needing to be mocked that wasn't needed before.

**Comment by MGibson1** [comment]
I'm willing to back down in light of greater opposition here, the main arguments for substitute were:

1. incumbant
2. matches server mocking paradigms.

1 is only relevant all other things being equal and 2 is becoming less of a positive and more of a why as we grow and specialize

**Comment by eliykat** [comment]
> The biggest difference in these are jest-mock-extended returns undefined for all methods of the mocked service. While @fluffy-spoon/substitute will return objects if the service returns an object. During my testing this showed up as some methods needing to be mocked that wasn't needed before.

Thanks for the heads up, this will probably require some tweaking of our existing tests. But in general it's not a problem. I personally prefer that mocks return `null` unless otherwise instructed, it seems like a more sensible default.

@Hinton @MGibson1 if you're okay with this, then we can merge this ADR and I'll create a tech debt ticket to track the work of deprecating Substitute from existing tests. Let me know if there's anything else you need from me, or any other concerns.

**Comment by Hinton** [comment]
I think the name of this ADR be a bit misleading since it's called `NSubstitute` in c# and you could become confused that we've deprecated the server mocking tool. Thoughts on calling it `clients-use-jest-mock.md` or something similar? (And update the title to reflect this)

File: decisions/0006-clients-use-jest-mocks.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+- It requires that we mix syntax, particularly for assertions: we use Jest's `expect` to assert the
+  test result, but we use Substitute's `Arg` to match arguments within mock calls.
+
+We should just pick one.
+
+## Considered Options
+
+- **Use Substitute and ban developers from using Jest mocks** - we keep both libraries and continue
+  to use them both in tandem. In this case, we should provide clear guidance and training to
+  developers to use Substitute for all mocking and argument matching (not Jest).
```
Comment (Hinton) [inline decisions/0006-clients-use-jest-mocks.md:26]: Eslint supports warning/erroring on imports which could help with this. But it's a bit clunky as the user might already have written all their tests using jest before finding out it's not allowed.

File: decisions/0006-clients-use-jest-mocks.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+- **Use a different mocking library and/or test runner** - we change to something else entirely.
+  This is not really on the table here, but it is an option. Jest is working well so far, so I don't
+  propose we change it.
+
+## Decision Outcome
+
+Chosen option: **Deprecate Substitute and use Jest with jest-mock-extended instead**
+
+We should also host a training/learning session to encourage and empower developers to unit test
+their code.
```
Comment (Hinton) [inline decisions/0006-clients-use-jest-mocks.md:43]: Sounds like a good opportunity for a Bitwarden Back talk? You interested?

**Review by Hinton — APPROVED** [review]
I'm happy with this. I ran the linter to make it a bit more readable.

**Comment by MGibson1** [comment]
@eliykat Are you OK with being assigned a tech debt to swap substitute with jest-mock-extended?

**Comment by eliykat** [comment]
@MGibson1 yes, I was going to use EC-319 to track both this ADR and the associated tech debt work.

**Pull request merged**